freeipa/src/main/resources/freeipa-salt/salt/freeipa/scripts/initdnarange.py

```diff
@@ -0,0 +1,156 @@
+from ipalib import api
+from ipaserver.install import replication
+import logging
+from random import randint
+from time import sleep
+
+
+class RangeSet:
+    def __init__(self, elements):
+        self.ranges = list(elements)
+
+    def __iter__(self):
+        return iter(self.ranges)
+
+    def __repr__(self):
+        return 'RangeSet: %r' % self.ranges
+
+    def has(self, tup):
+        for pos, i in enumerate(self.ranges):
+            if i[0] <= tup[0] and i[1] >= tup[1]:
+                return pos, i
+        raise ValueError('Invalid range or overlapping range')
+
+    def minus(self, tup):
+        pos, (x, y) = self.has(tup)
+        out = []
+        if x < tup[0]:
+            out.append((x, tup[0] - 1))
+        if y > tup[1]:
+            out.append((tup[1] + 1, y))
+        self.ranges[pos:pos + 1] = out
+
+    def __sub__(self, r):
+        r1 = RangeSet(self)
+        for i in r: r1.minus(i)
+        return r1
+
+    def sub(self, r):  # inplace subtraction
+        for i in r:
+            self.minus(i)
+
+    def subWithResult(self, result, r):  # unfortunately ipa console won't work with the __sub__ solution, couldn't initialize RangeSet
+        result.ranges = self.ranges
+        for i in r:
+            result.minus(i)
+        return result
+
+
+class DnaRangeIniter:
+    """Needed a class as FreeIPA console couldn't resolved method to method calls"""
+
+    MIN_RANGE_SIZE = 1000
+
+    def __init__(self, repl):
+        """Console couldn't resolve the import otherwise in class methods"""
+        self.replication = repl
+        self.localReplManager = self.replication.ReplicationManager(api.env.realm, api.env.host, starttls=True, port=389)
+
+    def validate_dna_range_missing(self):
+        dna_range_start, dna_range_end = self.localReplManager.get_DNA_range(api.env.host)
+        print ("DNA range on this server: {} - {}".format(dna_range_start, dna_range_end))
+        if dna_range_start is not None:
+            print ("DNA range is already set on this server, exiting")
+            exit(0)
+
+        dna_next_range_start, dna_next_range_end = self.localReplManager.get_DNA_next_range(api.env.host)
+        print ("DNA next range on this server: {} - {}".format(dna_next_range_start, dna_next_range_end))
+        if dna_next_range_start is not None:
+            print ("DNA next range is already set on this server, exiting")
+            exit(0)
+
+    def get_all_servers_fqdn(self):
+        servers_resp = api.Command.server_find(None, pkey_only=True)['result']
```

Comment:
I am not sure this handles ipa servers that were previously deleted. Is it possible that they the original ranges are reused?

Reply:
the previously deleted FreeIPA instances' ranges won't show up when the ranges are listed. So they are free to take and they can be reassigned. I tested the scenario having 2 freeipa instance, A and B. Created users on A, repaired it, so now it's C. C got the ranges from A and it continued to work, using the next id from the range.